### DISCUSSION TITLE
Newbie - Optimal Shift Pattern

### DISCUSSION THREAD

**Discussion by SteveHaine** [google/or-tools#2324] [CP-SAT questions] [answered]
OK so I'm new to or-tools an I'm interested if it can find an optimal shift pattern for a set of work. So I have various work to be done which start at different time and have different lengths, I need to find the most optimal (smallest number of employees) shift pattern to fulfil this demand, but as always their are constraints. 

Having done a bit of reading I guess this problem lends itself to CP-SAT. Ive seen the nurse scheduling problem but we need to be more dynamic i.e. some shifts can be 5hs and some can be 10hrs but at the end of the pattern each shift rotation must have done the same amount of work.

Any pointers would be gratefully received.

**Comment by Mizux** [comment]
Did you take a look at the [flexible] jobshop problem instead ?
and replace "machine" by "worker"...

src: https://github.com/google/or-tools/blob/stable/examples/python/flexible_job_shop_sat.py

**Reply by SteveHaine** [reply]
Thanks @Mizux, That Looks much closer to what's required, am I correct I saying the main area I would need to amend would be:

https://github.com/google/or-tools/blob/b77bd3ac69b7f3bb02f55b7bab6cbb4bab3917f2/examples/python/flexible_job_shop_sat.py#L148-L152

To add employee working time constraints?

Also how would it give me the minimum number of Machines (workers) required? 

Steve

**Reply by Mizux** [reply]
The objective here is to finish ASAP i.e. minimize makespan, so it won't minimize the number of workers.
https://github.com/google/or-tools/blob/b77bd3ac69b7f3bb02f55b7bab6cbb4bab3917f2/examples/python/flexible_job_shop_sat.py#L154-L157

You could try instead to create a boolean for each machine which will be true if at least one task is performed on it, then as objective minimize the sum of the boolean values...

**Comment by lperron** [comment, marked as answer]
maximize the sum of boolean values.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00
